eth/vm/rstack.py

```diff
@@ -0,0 +1,70 @@
+from typing import (
+    List,
+)
+
+from eth.exceptions import (
+    InsufficientStack,
+    FullStack,
+)
+
+from eth.validation import (
+    validate_stack_int,
+)
+
+from eth.utils import (
+    big_endian_to_int,
+    ValidationError,
+)
+
+
+"""
+This module simply implements for the return stack the exact same design used for the data stack.
+As this stack must simply push_int or pop1_int any time a subroutine is accessed or left, only those two functions are provided.
+For the same reason, the class RStack doesn't inherit from the abc StackAPI, as it would require to implement all the abstract methods defined.
+"""
+
+
+def _busted_type(item_type: type, value: Union[int, bytes]) -> ValidationError:
+    return ValidationError(
+        "Stack must always be bytes or int, "
+        f"got {item_type!r} type, val {value!r}"
+    )
+
+class RStack():
+    """
+    VM Return Stack
+    """
+
+    __slots__ = ['values', '_append', '_pop_int', '__len__']
+
+    def __init__(self) -> None:
+        values: List[int]
+        self.values = values
+        self._append = values.append
+        self._pop_typed = values.pop
+        self.__len__ = values.__len__
```

Comment:
You might consider using [`collections.UserList`](https://docs.python.org/3/library/collections.html#collections.UserList) for this class as I believe it comes built-in with some of the functionality you're doing here.

Reply:
Though perhaps it's maybe just as effective to subclass list, as the docs seem to recommend:
> The need for this class has been partially supplanted by the ability to subclass directly from list; however, this class can be easier to work with because the underlying list is accessible as an attribute.

Not sure there's much benefit to being able to access the underlying list via `.data` here, though.

---

Actually, on second thought, we don't really want/need to expose all the `list` methods anyway, so maybe the current approach is better.